### ISSUE TITLE
JSONEncoder uses _dump_date from werkzeug.http which doesn't format dates with years < 1000 correctly

### ISSUE THREAD

**Issue by nicolary** [pallets/flask#3514]
### Expected Behavior
Default date formatting converts dates with a year below 1000 to a non-zero padded string.
See _dump_date: https://github.com/pallets/werkzeug/blob/master/src/werkzeug/http.py

```
datetime.date(year=1, month=1, day=1)

should yield in the JSON: 'Mon, 01 Jan 0001 00:00:00 GMT'

so it can be parsed by:

'%a, %d %b %Y %H:%M:%S %Z'
```

### Actual Behavior
```
datetime.date(year=1, month=1, day=1)

yields in the JSON: 'Mon, 01 Jan 1 00:00:00 GMT'

which can't be parsed by:

'%a, %d %b %Y %H:%M:%S %Z'
```

### Environment

* Python version: 3.6.9
* Flask version: 1.1.1
* Werkzeug version: 0.16.1

### Comment

werkzeug may not want to apply the patch to _dump_date so maybe you want to implement your own safe http_date.

https://github.com/pallets/werkzeug/issues/1739

**Comment by davidism** [comment]
This is a Werkzeug issue, not Flask. I'm not clear why you closed the Werkzeug issue, I said I was open to a PR. I also pointed out that since you had specialized needs that are not typical of HTTP messages, you probably should use some custom serialization anyway.

**Comment by nicolary** [comment]
Sorry - I misunderstood. I didn't realize it was the same team working on both projects or I would have replied on the other ticket. I figured that the solution might be better if Flask didn't use http_date for generic datetime/date JSON serialization and the potentially best solution was to change Flask to use a less context bound function. I saw your point that _dump_date behind http_date was for a use case where dates before year 1000 weren't really relevant so I switched over to here to simply report it in case someone wanted to replace that http_date call with something more inclusive of all dates. I'll reply to the other ticket from now on.
